Dockerfile

```diff
@@ -2,45 +2,62 @@
 #
 # To run a container:
 #
-#   docker run -v /srv/portier-broker:/data:ro portier/broker /data/config.toml
+#   docker run --rm -v $(pwd):/data:ro portier/broker /data/config.toml
+#
+# To use your own data files (helpful in a production deployment):
+#
+#   tar zvcf data.tar.gz --exclude='*.po' lang res tmpl
+#   [copy data.tar.gz up to an HTTP server]
+#   docker build -t portier/broker:latest --build-arg=data=url --build-arg=data_url=https://example.com/data.tar.gz .
+#   docker run --rm -e ... portier/broker
 #
 
+# trick shamelessly stolen from https://stackoverflow.com/a/54245466
+ARG data=default
+
 # Create a release build.
-FROM rust:1-buster as build
+FROM rust:1-slim-buster AS build
+SHELL ["/bin/sh", "-x", "-c"]
+RUN apt-get update \
+  && apt-get -y --option=Dpkg::options::=--force-unsafe-io upgrade --no-install-recommends \
+  && apt-get -y --option=Dpkg::options::=--force-unsafe-io install --no-install-recommends \
+    libssl-dev \
+    pkg-config \
+  && apt-get clean \
+  && find /var/lib/apt/lists -type f -delete
 WORKDIR /src
 COPY . .
 RUN cargo build --release
-
-# Prepare a 'package' directory with the exact files we want.
-RUN set -x \
-  && mkdir package \
-  && cp -R \
-    lang \
-    res \
-    tmpl \
-    target/release/portier-broker \
-    package/ \
-  && rm lang/*.po
+RUN tar zcf /data.tar.gz --exclude='*.po' lang res tmpl
 
 # Prepare a final image from a plain Debian base.
-FROM debian:buster
+FROM debian:buster-slim AS data_default
+ONBUILD COPY --from=build /data.tar.gz /
+
+FROM debian:buster-slim AS data_url
+ONBUILD ARG data_url
+ONBUILD ADD ${data_url} /data.tar.gz
+
+FROM data_${data}
+
+SHELL ["/bin/sh", "-x", "-c"]
 
 # Add a user and group first to make sure their IDs get assigned consistently,
 # regardless of whatever dependencies get added.
-RUN set -x \
-  && groupadd -r -g 999 portier-broker \
+RUN groupadd -r -g 999 portier-broker \
   && useradd -r -g portier-broker -u 999 portier-broker
 
 # Install run-time dependencies.
-RUN set -x \
-  && apt-get update \
-  && apt-get install -y --no-install-recommends \
-    openssl \
+RUN apt-get update \
+  && apt-get -y --option=Dpkg::options::=--force-unsafe-io upgrade --no-install-recommends \
+  && apt-get -y --option=Dpkg::options::=--force-unsafe-io install --no-install-recommends \
     ca-certificates \
-  && rm -rf /var/lib/apt/lists/*
+  && apt-get clean \
+  && find /var/lib/apt/lists -type f -delete
 
 # Copy in the 'package' directory from the build image.
-COPY --from=build /src/package /opt/portier-broker
+RUN mkdir -p /opt/portier-broker && tar -C /opt/portier-broker -xf /data.tar.gz && rm /data.tar.gz
```

Comment:
I like this functionality, but wonder if there's not a simpler way to accomplish this. Do you think this could work:

1. Build stage
   - FROM debian
   - Build the broker
2. Data stage
   - FROM debian
   - Copy in lang, res, tmpl
   - If $data_url is set, rm the above, use curl + tar
   - rm *.po
3. Final stage
   - FROM debian
   - Setup users, install deps
   - Copy in broker from build stage
   - Copy in data from data stage
   - Image settings stuff

What I'm hoping to accomplish here:
 - In the final stage, I originally tried to keep deletions to a minimum. The final stage is what people download, and with your changes, there's a hidden data.tar.gz download in an upper layer, which a lower layer deletes. (This could be a petty amount of bytes, then again, someone could slam their data file full with branding. 🙃)
 - I believe changing data also reruns apt? Because the order makes Docker think apt commands have a dependance on data.
 - I think we can just be crude in stage 2, and don't need a switch. (Though the switch IS neat. 😄) Even if we copy in local data, then delete it again, to then rebuild it from $data_url, it doesn't matter because only the final stage dictates what layers are in the output. Only downside is we can't do a conditional `ADD`, so that's why curl + tar.

Reply:
Sure this works for me, not really bothered about how it gets done, just that the result is I do not need to fork the project to just add my own branding.

You want me to respin the Dockerfile along these lines or did you want to? I am not opinioned on the implementation side so if you are more passionate about it, you will not get any complaints from me!